pyanaconda/modules/subscription/system_purpose.py

```diff
@@ -0,0 +1,96 @@
+# 02110-1301, USA.  Any Red Hat trademarks that are incorporated in the
+# source code or documentation are not subject to the GNU General Public
+# License and may only be used or replicated with the express permission of
+# Red Hat, Inc.
+#
+
+import os
+import json
+
+from pyanaconda.core import util
+from pyanaconda.anaconda_loggers import get_module_logger
+log = get_module_logger(__name__)
+
+VALID_FIELDS_FILE_PATH = "/etc/rhsm/syspurpose/valid_fields.json"
+
+
+def get_valid_fields(valid_fields_file_path=VALID_FIELDS_FILE_PATH):
+    """Get valid role, sla and usage fields for system purpose use.
+
+    If no valid fields are provided, the fields file is not found or can not be
+    parsed then all three lists will be empty.
+
+    :param str valid_fields_file_path: path to a JSON file holding the valid field listings
+
+    :return: role, sla and usage type lists
+    :rtype: [roles], [slas], [usages]
+    """
+    valid_roles = []
+    valid_slas = []
+    valid_usage_types = []
+    if os.path.exists(valid_fields_file_path):
+        try:
+            with open(valid_fields_file_path, "rt") as f:
+                valid_fields_json = json.load(f)
+                valid_roles = valid_fields_json.get("role", [])
+                valid_slas = valid_fields_json.get("service_level_agreement", [])
+                valid_usage_types = valid_fields_json.get("usage", [])
+        except (IOError, json.JSONDecodeError):
+            log.exception("parsing of syspurpose valid fields file at %s failed",
+                          valid_fields_file_path)
+    else:
+        log.warning("system purpose valid fields file not found at %s", valid_fields_file_path)
+    return valid_roles, valid_slas, valid_usage_types
+
+
+def normalize_field(raw_field):
+    """Normalize a field for matching.
+
+    Fields specified in free form by users can have different case or trailing white space,
+    while still technically being a match on a valid field.
+
+    So convert the field to lower case and strip any trailing white space and return the result.
+
+    :param str raw_field: raw not normalized field
+    :return: normalized field suitable for matching
+    :rtype: str
+    """
+    return raw_field.strip().lower()
+
+
+def match_field(raw_field, valid_fields):
+    """Try to match the field on an item in a list of fields.
+
+    If a match is found return the first matching item from the list.
+    If np match is found, return None.
+
+    :param raw_field str: field to match
+    :param list valid_fields: list of valid fields to match against
+    :return: a matching valid field or None if no match is found
+    "rtype: str or None
+    """
+    matching_valid_field = None
+    normalized_field = normalize_field(raw_field)
+
+    for valid_field in valid_fields:
+        if normalized_field == normalize_field(valid_field):
+            # looks like the fields match, no need to search
+            # any further
```

Comment:
You can place this comment on one line. I don't think it will exceed 99 characters limit.

Reply:
You newer know! :D Changed. :)